Remove GetCallingAssembly usage

davidfowl opened the pull request:
- Today the WebApplicationBuilder tries to guess what the application name should be based on the immediate caller of the API using Assembly.GetCallingAssembly. This is because several components use the application name as a way to find the application assembly in order to do other things;
- User secrets uses it to find the location of the  directory that holds configuration secrets.
- MVC uses it as the assembly to start the search for controllers, pages, etc.
- Identifty uses it to find application parts.
This API is a bit fragile for internal refactoring (if the immediate caller is anything except for user code, then we will treat our own assembly as the calling assembly) and more importantly, it breaks in NativeAOT scenarios. Instead of the calling assembly, we now default to the entry assembly.
- This will "break" in scenarios where the WebApplication is started in a library and the entry point is the host application (controllers will be scanned in potentially the wrong place by default or it will find the wrong controllers).
- Now that we have the option that allows specifying the name of the application name via WebApplicationOptions, In the future, we're investigating ways that the compiler can inject the calling assembly via an attribute similar to [CallerMemberName] that would capture the calling assembly in a more reliable and AOT friendly way.
- Added a test that asserts that the default application name is the entry assembly.
- Updated the slnf file

Fixes #34838

davidfowl committed:
Remove GetCallingAssembly usage
- Today the WebApplicationBuilder tries to guess what the application name should be based on the immediate caller of the API using Assembly.GetCallingAssembly. This is because several components use the application name as a way to find the application assembly in order to do other things;
- User secrets uses it to find the location of the  directory that holds configuration secrets.
- MVC uses it as the assembly to start the search for controllers, pages, etc.
- Identifty uses it to find application parts.
This API is a bit fragile for internal refactoring (if the immediate caller is anything except for user code, then we will treat our own assembly as the calling assembly) and more importantly, it breaks in NativeAOT scenarios. Instead of the calling assembly, we now default to the entry assembly.
- This will "break" in scenarios where the WebApplication is started in a library and the entry point is the host application (controllers will be scanned in potentially the wrong place by default or it will find the wrong controllers).
- Now that we have the option that allows specifying the name of the application name via WebApplicationOptions, In the future, we're investigating ways that the compiler can inject the calling assembly via an attribute similar to [CallerMemberName] that would capture the calling assembly in a more reliable and AOT friendly way.
- Added a test that asserts that the default application name is the entry assembly.
- Updated the slnf file

davidfowl commented:
/azp run

captainsafia commented on src/DefaultBuilder/src/WebApplicationBuilder.cs:
```diff
@@ -61,7 +61,7 @@ internal WebApplicationBuilder(Assembly? callingAssembly, WebApplicationOptions
                 // We need to override the application name since the call to Configure will set it to
                 // be the calling assembly's name.
-                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, (callingAssembly ?? Assembly.GetEntryAssembly())?.GetName()?.Name ?? string.Empty);
+                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, (Assembly.GetEntryAssembly())?.GetName()?.Name ?? string.Empty);
```
Do we want people to be able ot set the assembly name directly from the options for the controllers in libraries scenarios?

davidfowl replied:
Yes, this is just the default. The API I added (WebApplicationOptions) allows setting the name. Or you can set it via command line args https://github.com/dotnet/aspnetcore/blob/8b06da35ddbc17280f02c0f1dcfd5fac1a40ea72/src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs#L387.

PS: You just made me realize I should update this test to be a different assembly name, now that the entry assembly is the default!

davidfowl committed:
PR feedback
- Change the name of the overriding assembly since the entry assembly is the default.

halter73 commented on src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:
```diff
@@ -386,7 +386,7 @@ public void WebApplicationBuilderApplicationNameDefaultsToEntryAssembly()
         public void WebApplicationBuilderApplicationNameCanBeOverridden()
         {
-            var assemblyName = Assembly.GetEntryAssembly().GetName().Name;
+            var assemblyName = typeof(WebApplicationTests).Assembly.GetName().Name;
```
Can we just pick a random name so we don't accidently lose coverage in case we revert behavior or something?

davidfowl replied:
The only reason it *could* be random is because we're not loading the assembly.
